File: java/fizzbuzz-outside-in/src/test/java/FizzBuzzTest.java
import com.mscharhag.oleaster.runner.OleasterRunner;
import org.junit.runner.RunWith;

import static com.mscharhag.oleaster.runner.StaticRunnerSupport.describe;
import static com.mscharhag.oleaster.runner.StaticRunnerSupport.it;
import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.Matchers.anyInt;
import static org.mockito.Mockito.*;

@RunWith(OleasterRunner.class)
public class FizzBuzzTest {
    static {
        describe("Fizz Buzz Generator", () -> {
            describe("Outside-in", () -> {
                it("should check whether each rule applies to the input", () -> {
                    int input = aNumber();
                    Rule rule1 = aRule();
                    Rule rule2 = aRule();
                    FizzBuzzGenerator generator = new FizzBuzzGenerator(rule1, rule2);
                    generator.calculate(input);
                    verify(rule1).appliesTo(input);
                    verify(rule2).appliesTo(input);
                });
                
                it("should stop checking if one matching rule was found", () -> {
                    Rule rule1 = aMatchingRule();
                    Rule rule2 = aRule();
                    FizzBuzzGenerator generator = new FizzBuzzGenerator(rule1, rule2);
                    generator.calculate(aNumber());
                    verify(rule2, never()).appliesTo(anyInt());
                });
                
                it("should return the result of the matching rule", () -> {
                    Rule rule = aMatchingRule();
                    String text = aText();
                    when(rule.getResult()).thenReturn(text);
                    FizzBuzzGenerator generator = new FizzBuzzGenerator(rule);
                    assertThat(generator.calculate(aNumber())).isEqualTo(text);
                });
                
                it("should return the input itself when no rule is matching", () -> {
                   FizzBuzzGenerator generator = new FizzBuzzGenerator(aRule());
                   assertThat(generator.calculate(23)).isEqualTo("23");
                });
            });
            
            it("should return expected values", () -> {
                FizzBuzzGenerator generator = new FizzBuzzGenerator();
                assertThat(generator.calculate(1)).isEqualTo("1");
                assertThat(generator.calculate(2)).isEqualTo("2");
                assertThat(generator.calculate(3)).isEqualTo("Fizz");
                assertThat(generator.calculate(5)).isEqualTo("Buzz");
                assertThat(generator.calculate(15)).isEqualTo("Fizz-Buzz");
            });
            
            it("zero should be returned as zero", () -> {
               FizzBuzzGenerator generator = new FizzBuzzGenerator();
               assertThat(generator.calculate(0)).isEqualTo("0");
            });
        });
    }

    private static String aText() {
        return "Something";
    }

    private static Rule aMatchingRule() {
        Rule rule = aRule();
        when(rule.appliesTo(anyInt())).thenReturn(true);
        return rule;
    }

    private static int aNumber() {
        return 42;
    }

    private static Rule aRule() {
        Rule mock = mock(Rule.class);
        when(mock.getResult()).thenReturn("");
        return mock;
    }
    
}

File: kotlin/fizzbuzz-outside-in/src/test/kotlin/FizzBuzzTest.kt
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Nested
import org.junit.jupiter.api.Test
import org.mockito.Mockito.*
import java.util.*

class FizzBuzzTest {
    @Nested
    inner class Algorithm {
        @Test
        fun `return input itself when no rule matches`() {
            val input = aNumber()
            val fizzBuzz = FizzBuzz(aRule(), aRule())
            assertThat(fizzBuzz.calculate(input)).isEqualTo(input.toString())
        }

        @Test
        fun `check if each rule applies to the input`() {
            val rule1 = aRule()
            val rule2 = aRule()
            val fizzBuzz = FizzBuzz(rule1, rule2)
            val input = aNumber()
            fizzBuzz.calculate(input)
            verify(rule1).appliesTo(input)
            verify(rule2).appliesTo(input)
        }

        @Test
        fun `return the result of the matching rule`() {
            val rule = aMatchingRule("Result")
            val fizzBuzz = FizzBuzz(rule)
            assertThat(fizzBuzz.calculate(aNumber())).isEqualTo("Result")
        }

        @Test
        fun `return the combination of multiple matching rules`() {
            val rule1 = aMatchingRule("Result1")
            val rule2 = aMatchingRule("Result2")
            val fizzBuzz = FizzBuzz(rule1, rule2)
            assertThat(fizzBuzz.calculate(aNumber())).isEqualTo("Result1-Result2")
        }

        private fun aMatchingRule(result: String): Rule {
            val rule = aRule()
            `when`(rule.appliesTo(anyInt())).thenReturn(true)
            `when`(rule.result).thenReturn(result)
            return rule
        }

        private fun aRule() = mock(Rule::class.java)

        private fun aNumber() = Math.abs(Random().nextInt(1000)) + 1
    }
    
    @Nested
    inner class ConcreteRules {
        @Test
        fun `return the expected values`() {
            val fizzBuzz = FizzBuzz()
            assertThat(fizzBuzz.calculate(1)).isEqualTo("1")
            assertThat(fizzBuzz.calculate(3)).isEqualTo("Fizz")
            assertThat(fizzBuzz.calculate(6)).isEqualTo("Fizz")
            assertThat(fizzBuzz.calculate(5)).isEqualTo("Buzz")
            assertThat(fizzBuzz.calculate(10)).isEqualTo("Buzz")
            assertThat(fizzBuzz.calculate(15)).isEqualTo("Fizz-Buzz")
        }
    }
}

